triage/summarize/summarize.go

```diff
@@ -0,0 +1,724 @@
+						globalClustering[other][testName] = make([]failure, len(tests))
+					}
+
+					// Copy the contents into the test's failure slice
+					globalClustering[other][testName] = append(globalClustering[other][testName], tests...)
+				} else {
+					clusterCase = "NEW"
+					globalClustering[key] = map[string][]failure{testName: tests}
+				}
+			}
+
+			clusterDuration := int(time.Since(clusterStart).Seconds())
+			log.Printf("  %4d/%4d clusters, %5d chars failure text, %5d failures, cluster:%s in %d sec, test: %s",
+				m, numClusters, fTextLen, numTests, clusterCase, clusterDuration, testName)
+	}
+	}
+
+	// If we seeded clusters using the previous run's keys, some of those
+	// clusters may have disappeared. Remove the resulting empty entries.
+	for key, val := range globalClustering {
+		if len(val) == 0 {
+			delete(globalClustering, key)
+		}
+	}
+
+	elapsed := time.Since(start)
+	log.Printf("Finished clustering %d unique tests (%d failures) into %d clusters in %s",
+		len(newlyClustered), numFailures, len(globalClustering), elapsed.String())
+
+	return globalClustering
+}
+
+/*
+testsGroupByJob takes a group of failures and a group of builds and returns the list of builds
+that belong to each job.
+
+builds is a mapping from build names to build objects.
+*/
+func tests_group_by_job(failures []failure, builds map[string]build) []struct {
+	jobName      string
+	buildNumbers []int
+} {
+	// groups maps job names to sets of build numbers.
+	var groups map[string]sets.Int
+
+	for _, flr := range failures {
+		var bld build
+		// Try to grab the build from builds if it exists
+		if matchedBuild, ok := builds[flr.build]; ok {
+			bld = matchedBuild
+		} else {
+			continue
+		}
+
+		// If the JSON build's "number" field was not null
+		if bld.number != 0 {
+			// Create the "set" if one doesn't exist for the given job
+			if _, ok := groups[bld.job]; !ok {
+				groups[bld.job] = make(sets.Int, 1)
+			}
+
+			groups[bld.job].Insert(bld.number)
+		}
+	}
+
+	// Sort groups in two stages
+
+	// First, sort each build number set in descending order
+	sortedBuildNumbers := make(map[string][]int, len(groups))
+	// Create the slice to hold the set elements, fill it, and sort it
+	for jobName, buildNumberSet := range groups {
+		// Initialize the int slice
+		sortedBuildNumbers[jobName] = make([]int, len(buildNumberSet))
+
+		// Fill it
+		iter := 0
+		for buildNumber := range buildNumberSet {
+			sortedBuildNumbers[jobName][iter] = buildNumber
+			iter++
+		}
+
+		// Sort it. Use > instead of < in less function to sort descending.
+		sort.Slice(sortedBuildNumbers[jobName], func(i, j int) bool { return sortedBuildNumbers[jobName][i] > sortedBuildNumbers[jobName][j] })
+	}
+
+	// Then, sort the jobs by the number of build numbers in each job's build number slice, descending.
+	// The struct acts as a tuple.
+	sortedGroups := make([]struct {
+		jobName      string
+		buildNumbers []int
+	}, len(groups))
```

Comment:
defining the struct twice feels like a sign it should be defined as a type outside of this func

Reply:
Done in d90c025